File: apps/web/src/services/backtest.ts
/**
 * Backtest API 服务
 */
import { api } from './api'

// ============================================
// 类型定义
// ============================================

export interface BacktestParams {
  start_date: string
  end_date: string
}

export interface BacktestJob {
  id: number
  agent_id: number
  status: 'pending' | 'running' | 'completed' | 'failed'
  progress: number
  params: BacktestParams
  created_at: string
  started_at?: string
  completed_at?: string
  error_msg?: string
}

export interface BacktestDetailItem {
  stock_code: string
  stock_name: string
  decision: string
  score: number
  reasoning?: string
  actual_hit: boolean
}

export interface BacktestDayResult {
  date: string
  total_signals: number
  hit_count: number
  miss_count: number
  hit_rate: number
  details: BacktestDetailItem[]
}

export interface BacktestResult {
  total_days: number
  total_signals: number
  total_hit: number
  total_miss: number
  overall_hit_rate: number
  days: BacktestDayResult[]
}

export interface BacktestResponse {
  id: number
  agent_id: number
  agent_name?: string
  status: string
  progress: number
  params: BacktestParams
  created_at: string
  started_at?: string
  completed_at?: string
  error_msg?: string
  result?: BacktestResult
}

export interface BacktestProgressResponse {
  id: number
  status: string
  progress: number
  message: string
}

export interface BacktestListResponse {
  items: BacktestJob[]
  total: number
}

export interface CreateBacktestRequest {
  agent_id: number
  start_date: string
  end_date: string
}

// ============================================
// API 方法
// ============================================

export const backtestApi = {
  // 创建回测任务
  createBacktest: async (data: CreateBacktestRequest): Promise<BacktestJob> => {
    const response = await api.post('/backtests', data)
    return response.data
  },

  // 获取回测列表
  getBacktests: async (params: { limit?: number; offset?: number } = {}): Promise<BacktestListResponse> => {
    const response = await api.get('/backtests', { params })
    return response.data
  },

  // 获取回测详情
  getBacktest: async (id: number): Promise<BacktestResponse> => {
    const response = await api.get(`/backtests/${id}`)
    return response.data
  },

  // 获取回测进度（轻量轮询）
  getBacktestProgress: async (id: number): Promise<BacktestProgressResponse> => {
    const response = await api.get(`/backtests/${id}/progress`)
    return response.data
  },
}

File: apps/api/internal/model/backtest.go
package model

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

// BacktestJob 回测任务表
type BacktestJob struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`

	UserID   uuid.UUID `json:"user_id" gorm:"type:uuid;index;not null"`
	AgentID  uint      `json:"agent_id" gorm:"index"`
	AgentType string   `json:"agent_type" gorm:"size:30;index"`
	Status   string    `json:"status" gorm:"size:20;default:'pending'"` // pending/running/completed/failed
	Params   JSON      `json:"params" gorm:"type:jsonb"`
	Progress int       `json:"progress" gorm:"default:0"`
	ErrorMsg string    `json:"error_msg,omitempty" gorm:"type:text"`

	StartedAt   *time.Time `json:"started_at,omitempty"`
	CompletedAt *time.Time `json:"completed_at,omitempty"`
}

// TableName 指定表名
func (BacktestJob) TableName() string {
	return "backtest_jobs"
}

// BacktestParams 回测参数
 type BacktestParams struct {
	StartDate string `json:"start_date"`
	EndDate   string `json:"end_date"`
}

// BacktestResult 回测结果汇总表
type BacktestResult struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`

	JobID           uint    `json:"job_id" gorm:"unique;not null"`
	AgentID         uint    `json:"agent_id" gorm:"index"`
	TotalDays       int     `json:"total_days" gorm:"not null;default:0"`
	TotalSignals    int     `json:"total_signals" gorm:"not null;default:0"`
	TotalHit        int     `json:"total_hit" gorm:"not null;default:0"`
	TotalMiss       int     `json:"total_miss" gorm:"not null;default:0"`
	OverallHitRate  float64 `json:"overall_hit_rate" gorm:"type:decimal(5,4);default:0"`
}

// TableName 指定表名
func (BacktestResult) TableName() string {
	return "backtest_results"
}

// BacktestDayResult 回测按天结果表
type BacktestDayResult struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	CreatedAt time.Time `json:"created_at"`

	ResultID     uint      `json:"result_id" gorm:"index;not null"`
	Date         time.Time `json:"date" gorm:"type:date;not null"`
	TotalSignals int       `json:"total_signals" gorm:"not null;default:0"`
	HitCount     int       `json:"hit_count" gorm:"not null;default:0"`
	MissCount    int       `json:"miss_count" gorm:"not null;default:0"`
	HitRate      float64         `json:"hit_rate" gorm:"type:decimal(5,4);default:0"`
	Details      json.RawMessage `json:"details" gorm:"type:jsonb"`
}

// TableName 指定表名
func (BacktestDayResult) TableName() string {
	return "backtest_day_results"
}

// BacktestDetailItem 回测详情单项（存储在 BacktestDayResult.Details JSON 中）
type BacktestDetailItem struct {
	StockCode string  `json:"stock_code"`
	StockName string  `json:"stock_name"`
	Decision  string  `json:"decision"`
	Score     float64 `json:"score,omitempty"`
	Reasoning string  `json:"reasoning,omitempty"`
	ActualHit bool    `json:"actual_hit"`
}
